golem-component-service-base/src/repo/component.rs

```diff
@@ -0,0 +1,467 @@
+        Ok(Self {
+            namespace: namespace.to_string(),
+            component_id: component.versioned_component_id.component_id.0,
+            name: component.component_name.0,
+            size: component.component_size as i32,
+            version: component.versioned_component_id.version as i64,
+            user_component: component.versioned_component_id.slug(),
+            protected_component: component.protected_component_id.slug(),
+            protector_version: None,
+            metadata: metadata.into(),
+        })
+    }
+}
+
+#[async_trait]
+pub trait ComponentRepo {
+    async fn upsert(&self, component: &ComponentRecord) -> Result<(), RepoError>;
+
+    async fn get(
+        &self,
+        namespace: &str,
+        component_id: &Uuid,
+    ) -> Result<Vec<ComponentRecord>, RepoError>;
+
+    async fn get_all(&self, namespace: &str) -> Result<Vec<ComponentRecord>, RepoError>;
+
+    async fn get_latest_version(
+        &self,
+        namespace: &str,
+        component_id: &Uuid,
+    ) -> Result<Option<ComponentRecord>, RepoError>;
+
+    async fn get_by_version(
+        &self,
+        namespace: &str,
+        component_id: &Uuid,
+        version: u64,
+    ) -> Result<Option<ComponentRecord>, RepoError>;
+
+    async fn get_by_name(
+        &self,
+        namespace: &str,
+        name: &str,
+    ) -> Result<Vec<ComponentRecord>, RepoError>;
+
+    async fn get_ids_by_name(&self, namespace: &str, name: &str) -> Result<Vec<Uuid>, RepoError>;
+
+    async fn get_namespaces(&self, component_id: &Uuid) -> Result<Vec<(String, i64)>, RepoError>;
+
+    async fn delete(&self, namespace: &str, component_id: &Uuid) -> Result<(), RepoError>;
+}
+
+pub struct DbComponentRepo<DB: Database> {
+    db_pool: Arc<Pool<DB>>,
+}
+
+impl<DB: Database> DbComponentRepo<DB> {
+    pub fn new(db_pool: Arc<Pool<DB>>) -> Self {
+        Self { db_pool }
+    }
+}
+
+#[async_trait]
+impl ComponentRepo for DbComponentRepo<sqlx::Sqlite> {
+    async fn upsert(&self, component: &ComponentRecord) -> Result<(), RepoError> {
+        let mut transaction = self.db_pool.begin().await?;
+
+        let result = sqlx::query(
+            "SELECT count(component_id) as count FROM components WHERE namespace != $1 AND component_id = $2",
+        )
+        .bind(component.namespace.clone())
+        .bind(component.component_id)
+        .fetch_one(&mut *transaction)
+        .await?;
+
+        let count: i64 = result.get("count");
+
+        if count > 0 {
+            return Err(RepoError::Internal("Component id not unique".to_string()));
+        }
+
+        sqlx::query(
+            r#"
+              INSERT INTO components
+                (namespace, component_id, version, name, size, user_component, protected_component, protector_version, metadata)
+              VALUES
+                ($1, $2, $3, $4, $5, $6, $7, $8, $9)
+              ON CONFLICT (namespace, component_id, version) DO UPDATE
+              SET name = $4,
```

Comment:
This implementation allows different names for different versions on the same `component_id`. The only solution without races is to normalize the schema.

Reply:
changed to insert only